File: src/game/level/Level.java
package game.level;
import game.Bullet;
import game.Vector;
import game.creature.Creature;
import game.creature.Player;
import game.creature.TurretZombie;
import game.item.Item;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;


public class Level {
	public String levelName;
	public TileGrid tileGrid;
	public List<Creature> creatures = new ArrayList<Creature>();
	public List<Creature> spawnReqs;
	public Player player;
	public Creature boss;
	public List<Item> items = new ArrayList<Item>();
	public List<Bullet> bullets = new ArrayList<Bullet>();
	public boolean didWin = false;
	public boolean didDie = false;
	
	
	public Level(String levelName, int numRows, int numCols) {
		this.levelName = levelName;
		tileGrid = new TileGrid(numRows, numCols);
		player = new Player();
		player.pos = Vector.zeroVector;
		player.angle = 0;
		creatures.add(player);
	}
	
	public Level(String levelName) throws IOException {
		this.levelName = levelName;
		String levelPath = "res/levels/" + levelName;
		
		tileGrid = TileGrid.loadFromFile(levelPath + ".tg");
		creatures = SpawnChart.loadFromFile(levelPath + ".sc").instantiate();
		items = ItemChart.loadFromFile(levelPath + ".ic").instantiate();
		
		for (Creature c : creatures) {
			if (c instanceof Player) {
				player = (Player) c;
				break;
			}
		}
		if (player == null) {
			player = new Player();
			player.pos = Vector.zeroVector;
			player.angle = 0;
			creatures.add(player);
		}
		
		if (levelName.equals("b")) {
			for (Creature c : creatures) {
				if (c instanceof TurretZombie) {
					boss = c;
					break;
				}
			}
		}
		
		if (levelName.equals("3") || levelName.equals("B")) player.ammo = 0;
	}
	
	
	public void update(float dt) {
		spawnReqs = new ArrayList<>();
		
		for (Creature c : creatures)
			c.update(dt, this);
		
		updateBullets(dt);
		checkCollisions();
		
		for (int i = 0; i < creatures.size(); i++) {
			if (creatures.get(i).health <= 0) {
				if (creatures.get(i) == player) {
					System.out.println("Player has died.");
					didDie = true;
				}
				
				creatures.remove(i);
				i--;
			}
		}
		
		updateItems(dt);
		
		if (player.hasKey) {
			int[] pTile = tileGrid.getTile(player.pos);
			if (pTile != null && pTile[0] == tileGrid.exitRow && pTile[1] == tileGrid.exitCol) {
				System.out.println("Level complete!");
				didWin = true;
			}
		}
		
		creatures.addAll(spawnReqs);
	}
	
	private void updateBullets(float dt) {
		for (int i = 0; i < bullets.size(); i++) {
			Bullet b = bullets.get(i);
			b.update(dt);
			
			boolean struck = false;
			for (Creature c : creatures) {
				if (b.collide(c)) {
					struck = true;
					break;
				}
			}
			if (!struck) {
				int[] tile = tileGrid.getTile(b.pos);
				if (tile != null && tileGrid.tiles[tile[0]][tile[1]])
					struck = true;
			}
			if (struck) {
				bullets.remove(i);
				i--;
			}
		}
	}
	
	private void updateItems(float dt) {
		boolean hadKey = player.hasKey;
		
		if (boss != null && boss.health <= 0) player.hasKey = true;
		
		for (int i = 0; i < items.size(); i++) {
			Item item = items.get(i);
			
			boolean pickedUp = false;
			for (Creature c : creatures) {
				if (!c.solid) continue;
				if (c.pos.distance(item.pos) <= c.hitRadius + item.radius) {
					if (item.pickedUpBy(c)) {
						pickedUp = true;
						break;
					}
				}
			}
			if (pickedUp) {
				items.remove(i);
				i--;
			}
		}
		
		if (!hadKey && player.hasKey) {
			for (int[] tile : tileGrid.keyActivated) {
				tileGrid.tiles[tile[0]][tile[1]] = !tileGrid.tiles[tile[0]][tile[1]];
			}
		}
	}
	
	private void checkCollisions() {
		for (int i = 0; i < creatures.size(); i++) {
			for (int j = i+1; j < creatures.size(); j++) {
				creatures.get(i).collide(creatures.get(j));
			}
		}
		for (Creature c : creatures)
			tileGrid.collide(c);
	}


	public void render() {
		tileGrid.render();
		
		for (Item i : items) {
			i.render();
		}
		for (Creature c : creatures) {
			c.render();
		}
		for (Bullet b : bullets) {
			b.render();
		}
	}
}

File: ts/game/level/Level.ts
import * as PIXI from 'pixi.js';
import { Bullet } from '../Bullet.js';
import { Vector } from '../Vector.js';
import { Creature } from '../creature/Creature.js';
import { Player } from '../creature/Player.js';
import { TurretZombie } from '../creature/TurretZombie.js';
import { Item } from '../item/Item.js';
import { TileGrid } from './TileGrid.js';
import { instantiateSpawnChart, loadSpawnChart } from './SpawnChart.js';
import { instantiateItemChart, loadItemChart } from './ItemChart.js';

export class Level {
  tileGrid: TileGrid;
  creatures: Creature[] = [];
  spawnReqs: Creature[] = [];
  player: Player;
  boss?: Creature;
  items: Item[] = [];
  bullets: Bullet[] = [];
  didWin = false;
  didDie = false;

  constructor(public levelName: string, tileGrid: TileGrid, creatures: Creature[], items: Item[]) {
    this.tileGrid = tileGrid;
    this.creatures = creatures;
    this.items = items;
    this.player = this.creatures.find(c => c instanceof Player) as Player;
    if (!this.player) {
      this.player = new Player();
      this.player.pos = Vector.zeroVector;
      this.player.angle = 0;
      this.creatures.push(this.player);
    }
    if (levelName === 'b') {
      for (const c of this.creatures) if (c instanceof TurretZombie) { this.boss = c; break; }
    }
    if (levelName === '3' || levelName === 'b') this.player.ammo = 0;
  }

  static async load(levelName: string): Promise<Level> {
    const levelPath = `res/levels/${levelName}`;
    const tileGrid = await TileGrid.loadFromFile(`${levelPath}.tg`);
    const creatures = instantiateSpawnChart(await loadSpawnChart(`${levelPath}.sc`));
    const items = instantiateItemChart(await loadItemChart(`${levelPath}.ic`));
    return new Level(levelName, tileGrid, creatures, items);
  }

  update(dt: number) {
    this.spawnReqs = [];
    for (const c of this.creatures) c.update(dt, this);
    this.updateBullets(dt);
    this.checkCollisions();
    for (let i = 0; i < this.creatures.length; i++) {
      if (this.creatures[i].health <= 0) {
        if (this.creatures[i] === this.player) {
          console.log('Player has died.');
          this.didDie = true;
        }
        const disp = this.creatures[i].display;
        if (disp && disp.parent) disp.parent.removeChild(disp);
        this.creatures.splice(i, 1);
        i--;
      }
    }
    this.updateItems(dt);
    if (this.player.hasKey) {
      const pTile = this.tileGrid.getTile(this.player.pos);
      if (pTile && pTile[0] === this.tileGrid.exitRow && pTile[1] === this.tileGrid.exitCol) {
        console.log('Level complete!');
        this.didWin = true;
      }
    }
    this.creatures.push(...this.spawnReqs);
  }

  private updateBullets(dt: number) {
    for (let i = 0; i < this.bullets.length; i++) {
      const b = this.bullets[i];
      b.update(dt);
      let struck = false;
      for (const c of this.creatures) {
        if (b.collide(c)) { struck = true; break; }
      }
      if (!struck) {
        const tile = this.tileGrid.getTile(b.pos);
        if (tile && this.tileGrid.tiles[tile[0]][tile[1]]) struck = true;
      }
      if (struck) {
        if (b.graphics.parent) b.graphics.parent.removeChild(b.graphics);
        this.bullets.splice(i, 1);
        i--;
      }
    }
  }

  private updateItems(_dt: number) {
    const hadKey = this.player.hasKey;
    if (this.boss && this.boss.health <= 0) this.player.hasKey = true;
    for (let i = 0; i < this.items.length; i++) {
      const item = this.items[i];
      let pickedUp = false;
      for (const c of this.creatures) {
        if (!c.solid) continue;
        if (c.pos.distance(item.pos) <= c.hitRadius + item.radius) {
          if (item.pickedUpBy(c)) { pickedUp = true; break; }
        }
      }
      if (pickedUp) {
        if (item.display && item.display.parent) {
          item.display.parent.removeChild(item.display);
        }
        this.items.splice(i, 1);
        i--;
      }
    }
    if (!hadKey && this.player.hasKey) {
      for (const tile of this.tileGrid.keyActivated) {
        this.tileGrid.tiles[tile[0]][tile[1]] = !this.tileGrid.tiles[tile[0]][tile[1]];
      }
    }
  }

  private checkCollisions() {
    for (let i = 0; i < this.creatures.length; i++) {
      for (let j = i + 1; j < this.creatures.length; j++) {
        this.creatures[i].collide(this.creatures[j]);
      }
    }
    for (const c of this.creatures) this.tileGrid.collide(c);
  }

  render(container: PIXI.Container) {
    this.tileGrid.render(container);
    for (const i of this.items) i.render(container);
    for (const c of this.creatures) c.render(container);
    for (const b of this.bullets) b.render(container);
    this.tileGrid.renderOverlays(container);
  }
}
